### ISSUE TITLE
Explicitly state when to use amp vs FP16_Optimizer in Documentation

### ISSUE THREAD

**Issue by FabianIsensee** [NVIDIA/apex#159]
Hi everyone,
first and foremost THANK YOU for this great respository! FP16 training was always a pain to get working properly and thanks to apex/amp it is now super easy!
When I read the documentation, one question was never answered specifically, but would have made my life as a user a lot easier if it were: What is exactly the difference between amp and FP16_Optimizer? What are the use cases of using FP16_Optimizer over amp or vise versa? Can they be used in conjunction?
I think adding a clear distinction between FP16_Optimizer that answers the questions form above would make it a lot easier for new users to find what they need.
(The current description is not very clear on this:

> The intention of amp is to be the "on-ramp" to easy FP16 training: achieve all the numerical stability of full FP32 training, with most of the performance benefits of full FP16 training.

> The intention of FP16_Optimizer is to be the "highway" for FP16 training: achieve most of the numerically stability of full FP32 training, and almost all the performance benefits of full FP16 training.

)
Best,
Fabian

**Comment by mcarilli** [comment]
Apex came together as a set of different tools from different sources that were similar enough that it made sense to put them in one place.  We were pleasantly surprised (and frankly I was a little overwhelmed) by how popular it became.  Right now my top priority is creating a unified interface for the different mixed precision tools.  This interface will make it much clearer what to use when, and also improve performance for all tools.  Branch [api_refactor](https://github.com/NVIDIA/apex/compare/api_refactor?expand=1) is tracking my changes.  I've got it working, but I don't have any documentation or examples, and it needs more stress testing.  I plan to merge it into master by next week.

I will be introducing the new API in a webinar tomorrow:  https://info.nvidia.com/webinar-mixed-precision-with-pytorch-reg-page.html.  This webinar should also clarify many of your questions.

The direct answer to your question is, FP16_Optimizer is designed to implement the "mixed precision recipe" (master weights and loss scaling) around an existing **half-precision** model.  FP16_Optimizer does not alter Torch internal functions.

Amp is a different approach to mixed precision:  It monkey-patches all torch functions and Tensor methods to cast input to the appropriate type (e.g. fp16 for Tensor Core-friendly ops like gemms and convolutions, and fp32 for operations that are safer with more precision, like softmax).  You do not have to (and should not) manually cast your data or model to half when using Amp.

Amp and FP16_Optimizer should not be used in conjunction.  Again, this will all become much clearer with the new API and I highly recommend you attend the webinar and stay tuned for the new API to be merged (within a week).

**Comment by FabianIsensee** [comment]
Thank you so much for the detailed explanation! This was very helpful! I already have amp running, but now I understand a bit better what is going on under the hood. I will definitely be there for the webinar :-)
Best,
Fabian
